pkg/boot/kexec/purg_load.go

```diff
@@ -14,36 +14,42 @@ import (
 	"github.com/u-root/u-root/pkg/boot/align"
 )
 
-// PurgeLoad loads an elf file at the desired address.
+// PurgeLoad loads an ELF (which must be one of the two purgatory from the
+// pkg/boot/linux package) into kmem, instructing it to jump to entry with RSI
+// set to rsi.
 func PurgeLoad(kmem *Memory, elfBuf []byte, start, param uintptr) (uintptr, error) {
 	elfFile, err := elf.NewFile(bytes.NewReader(elfBuf))
 	if err != nil {
-		return 0, fmt.Errorf("parse elf file from elf buffer: %v", err)
+		return 0, fmt.Errorf("parse purgatory ELF file from ELF buffer: %v", err)
 	}
 
 	log.Printf("Elf file: %#v, %d Progs", elfFile, len(elfFile.Progs))
 	if len(elfFile.Progs) != 1 {
-		return 0, fmt.Errorf("parse elf file: can only handle one Prog, not %d", len(elfFile.Progs))
+		return 0, fmt.Errorf("parse purgatory ELF file: can only handle one Prog, not %d", len(elfFile.Progs))
 	}
 	p := elfFile.Progs[0]
+
 	// the package really wants things page-sized, and rather than
 	// deal with all the bugs that arise from that, just keep it happy.
 	p.Memsz = uint64(align.AlignUpPageSize(uint(p.Memsz)))
 	b := make([]byte, p.Memsz)
 	if _, err := p.ReadAt(b[:p.Filesz], 0); err != nil {
 		return 0, err
 	}
-	entry := elfFile.Entry
+	entry := uintptr(elfFile.Entry)
 	Debug("Start is %#x, param is %#x", start, param)
+
 	binary.LittleEndian.PutUint64(b[8:], uint64(start))
 	binary.LittleEndian.PutUint64(b[16:], uint64(param))
+
+	// TODO: Shouldn't the purgatories be relocatable?
```

Comment:
no, the simplification comes from making them not relocatable. Removing the need to do a relink step at load time makes life a ton easier. We can iterate over the elf Programs, not the elf Sections, which we can not do with the standard purgatories -- they don't even have Program sections, since the final link step has not been done.

They can be written as position independent code, in future, however, and I'm fine with that. I just don't want to process ELF sections to produce a final binary in the u-root kexec.

Reply:
Agreed on no processing. The Go purgatory in pkg/kexec/multiboot/trampoline that's equivalent to the 32bit one you have in here is truly position independent and therefore relocatable. It's the most ideal position to be in rather than hard-coding a value.

Reply:
i can change the Todo to "position independent". I'm mentioning the other one to say: it's not that hard to make it position independent, we've done it before.

Reply:
done